source/volume_of_fluid/utilities.cc

```diff
@@ -0,0 +1,687 @@
+      {
+        AssertThrow(false, ExcNotImplemented());
+      }
+
+
+      template<int dim>
+      double compute_interface_location_newton(const int degree,
+                                               const Tensor<1, dim> normal,
+                                               const double volume_fraction,
+                                               const double vol,
+                                               const double epsilon,
+                                               const std::vector<Point<dim>> &points,
+                                               const std::vector<double> &weights)
+      {
+        double norm1=0.0;
+        for (int i=0; i<dim; ++i) norm1+=fabs(normal[i]);
+        double d_l=-0.5L*norm1, d_h=0.5L*norm1;
+        double f_l=0.0, f_h=1.0;
+        double d_guess= d_l + (volume_fraction-f_l)*(d_h-d_l)/(f_h-f_l);
+        double f_guess, df_guess;
+
+        std::vector<double> f_values(points.size());
+        std::vector<double> df_values(points.size());
+
+        for (int iter=0; iter<40; ++iter)
+          {
+            xFEM_Heaviside(degree, normal, d_guess, points, f_values);
+            xFEM_Heaviside_derivative_d(degree, normal, d_guess, points, df_values);
+
+            f_guess=0.0;
+            df_guess=0.0;
+            for (unsigned int i=0; i<points.size(); ++i)
+              {
+                const double factor = weights[i]/vol;
+                f_guess  += f_values[i]*factor;
+                df_guess += df_values[i]*factor;
+              }
+
+            // Break if within tolerance
+            if (fabs(f_guess-volume_fraction)<epsilon)
+              {
+                break;
+              }
+
+            if (volume_fraction<f_guess)
+              {
+                d_h = d_guess;
+                f_h = f_guess;
+              }
+            else
+              {
+                d_l = d_guess;
+                f_l = f_guess;
+              }
+
+            if (fabs(df_guess)<epsilon)
+              {
+                d_guess = (volume_fraction-f_l)/(f_h-f_l)*(d_h-d_l);
+              }
+            else
+              {
+                d_guess += (volume_fraction-f_guess)/(df_guess);
+
+                if (d_guess < d_l || d_guess > d_h)
+                  {
+                    d_guess = d_l + (volume_fraction-f_l)*(d_h-d_l)/(f_h-f_l);
+                  }
+              }
+          }
+
+        return d_guess;
+      }
+
+      template<int dim>
+      double compute_fluid_volume(const int degree,
+                                  const Tensor<1, dim> normal,
+                                  const double d,
+                                  const std::vector<Point<dim>> &points,
+                                  const std::vector<double> &weights)
+      {
+        std::vector<double> f_values(points.size());
+
+        xFEM_Heaviside(degree, normal, d, points, f_values);
+
+        double vol=0.0;
```

Comment:
Is this `volume_fraction`, or `cell_volume`? based on the loop below I suppose it is `volume_fraction`, but you used `vol` as abbreviation for `cell_volume` in the function above. This is an example for why we ask so persistent for self-explanatory variable names.